File: sdks/android/src/main/java/com/arbitrationsdk/android/models/ArbitrationModels.kt
package com.arbitrationsdk.android.models

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.Serializable
import java.util.*
import kotlin.time.Duration

/**
 * Analysis result containing all arbitration detection information
 */
@Parcelize
@Serializable
data class ArbitrationAnalysisResult(
    val id: String,
    val hasArbitration: Boolean,
    val confidence: Double,
    val keywordMatches: List<KeywordMatch>,
    val exclusionMatches: List<KeywordMatch>,
    val riskLevel: RiskLevel,
    val recommendations: List<String>,
    val metadata: AnalysisMetadata
) : Parcelable

/**
 * Keyword match information
 */
@Parcelize
@Serializable
data class KeywordMatch(
    val keyword: String,
    val range: String, // Stored as "start:end"
    val context: String,
    val confidence: Double
) : Parcelable

/**
 * Risk level enumeration
 */
@Parcelize
@Serializable
enum class RiskLevel(val displayName: String, val colorHex: String) : Parcelable {
    MINIMAL("Minimal Risk", "#4CAF50"),
    LOW("Low Risk", "#FFC107"),
    MEDIUM("Medium Risk", "#FF9800"),
    HIGH("High Risk", "#F44336");

    companion object {
        fun fromConfidence(confidence: Double): RiskLevel {
            return when {
                confidence >= 0.8 -> HIGH
                confidence >= 0.5 -> MEDIUM
                confidence >= 0.2 -> LOW
                else -> MINIMAL
            }
        }
    }
}

/**
 * Analysis metadata
 */
@Parcelize
@Serializable
data class AnalysisMetadata(
    val textLength: Int,
    @Serializable(with = DurationSerializer::class)
    val processingTime: Duration,
    val sdkVersion: String,
    val modelVersion: String,
    @Serializable(with = DateSerializer::class)
    val analysisDate: Date
) : Parcelable

/**
 * Rule-based analysis result
 */
data class RuleBasedResult(
    val hasArbitration: Boolean,
    val confidence: Double,
    val keywordMatches: List<KeywordMatch>,
    val exclusionMatches: List<KeywordMatch>
)

/**
 * Document type enumeration
 */
@Parcelize
@Serializable
enum class DocumentType(val extensions: List<String>) : Parcelable {
    PDF(listOf("pdf")),
    WORD(listOf("doc", "docx")),
    TEXT(listOf("txt")),
    RTF(listOf("rtf")),
    HTML(listOf("html", "htm")),
    IMAGE(listOf("jpg", "jpeg", "png", "tiff", "bmp"));

    companion object {
        fun fromExtension(extension: String): DocumentType? {
            return values().find { type ->
                type.extensions.contains(extension.lowercase())
            }
        }
    }
}

/**
 * SDK Configuration
 */
@Parcelize
@Serializable
data class SDKConfiguration(
    val apiBaseUrl: String = "https://api.arbitration-platform.com",
    val apiKey: String = "",
    val confidenceThreshold: Double = 0.5,
    val enableOfflineMode: Boolean = true,
    val enableAnalytics: Boolean = true,
    val cacheSize: Int = 100,
    val networkTimeoutSeconds: Int = 30
) : Parcelable {
    companion object {
        val DEFAULT = SDKConfiguration()
    }
}

/**
 * Arbitration error types
 */
sealed class ArbitrationError(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class InvalidInput(message: String) : ArbitrationError(message)
    class AnalysisFailure(message: String, cause: Throwable? = null) : ArbitrationError(message, cause)
    class NetworkError(message: String, cause: Throwable? = null) : ArbitrationError(message, cause)
    class StorageError(message: String, cause: Throwable? = null) : ArbitrationError(message, cause)
    class UnsupportedDocumentType(message: String) : ArbitrationError(message)
    class AuthenticationFailure(message: String) : ArbitrationError(message)
    object RateLimitExceeded : ArbitrationError("Rate limit exceeded. Please try again later.")
    object ModelNotAvailable : ArbitrationError("ML model is not available. Using fallback detection.")
}

/**
 * Analysis request
 */
@Parcelize
@Serializable
data class AnalysisRequest(
    val text: String,
    val requestId: String = UUID.randomUUID().toString(),
    val options: AnalysisOptions = AnalysisOptions.DEFAULT
) : Parcelable

/**
 * Analysis options
 */
@Parcelize
@Serializable
data class AnalysisOptions(
    val includeConfidenceScore: Boolean = true,
    val includeKeywordMatches: Boolean = true,
    val includeRecommendations: Boolean = true,
    val language: String = "en",
    val strictMode: Boolean = false
) : Parcelable {
    companion object {
        val DEFAULT = AnalysisOptions()
    }
}

/**
 * Batch analysis result
 */
@Parcelize
@Serializable
data class BatchAnalysisResult(
    val batchId: String,
    val results: List<ArbitrationAnalysisResult>,
    val summary: BatchSummary,
    @Serializable(with = DurationSerializer::class)
    val processingTime: Duration
) : Parcelable

/**
 * Batch summary
 */
@Parcelize
@Serializable
data class BatchSummary(
    val totalDocuments: Int,
    val documentsWithArbitration: Int,
    val averageConfidence: Double,
    val highRiskDocuments: Int,
    val processingErrors: Int
) : Parcelable

/**
 * Analytics event
 */
@Parcelize
@Serializable
data class AnalyticsEvent(
    val eventType: AnalyticsEventType,
    @Serializable(with = DateSerializer::class)
    val timestamp: Date = Date(),
    val properties: Map<String, String> = emptyMap()
) : Parcelable

/**
 * Analytics event types
 */
@Parcelize
@Serializable
enum class AnalyticsEventType : Parcelable {
    ANALYSIS_STARTED,
    ANALYSIS_COMPLETED,
    ANALYSIS_FAILED,
    DOCUMENT_SCANNED,
    OFFLINE_MODE_ENABLED,
    CACHE_HIT,
    API_CALL_MADE
}

/**
 * Notification settings
 */
@Parcelize
@Serializable
data class NotificationSettings(
    val enablePushNotifications: Boolean = true,
    val enableAnalysisComplete: Boolean = true,
    val enableHighRiskAlerts: Boolean = true,
    val enableWeeklyReports: Boolean = false
) : Parcelable

/**
 * User preferences
 */
@Parcelize
@Serializable
data class UserPreferences(
    val language: String = "en",
    val theme: AppTheme = AppTheme.SYSTEM,
    val notifications: NotificationSettings = NotificationSettings(),
    val autoSave: Boolean = true,
    val biometricAuth: Boolean = false
) : Parcelable

/**
 * App theme
 */
@Parcelize
@Serializable
enum class AppTheme(val displayName: String) : Parcelable {
    LIGHT("Light"),
    DARK("Dark"),
    SYSTEM("System")
}

/**
 * Storage-related models
 */

/**
 * Sort options for analysis results
 */
enum class SortOption(val displayName: String) {
    DATE_ASCENDING("Date (Oldest First)"),
    DATE_DESCENDING("Date (Newest First)"),
    CONFIDENCE_ASCENDING("Confidence (Low to High)"),
    CONFIDENCE_DESCENDING("Confidence (High to Low)")
}

/**
 * Result filter
 */
data class ResultFilter(
    val hasArbitration: Boolean? = null,
    val minConfidence: Double? = null,
    val maxConfidence: Double? = null,
    val riskLevel: RiskLevel? = null,
    val dateFrom: Date? = null,
    val dateTo: Date? = null
)

/**
 * Storage statistics
 */
data class StorageStatistics(
    val totalAnalyses: Int,
    val analysesWithArbitration: Int,
    val totalEvents: Int,
    val unsyncedEvents: Int,
    val cacheHitRate: Double,
    val databaseSizeBytes: Long
)

/**
 * Network-related models
 */

/**
 * Analysis history response
 */
@Serializable
data class AnalysisHistoryResponse(
    val results: List<ArbitrationAnalysisResult>,
    val pagination: PaginationInfo,
    val totalCount: Int
)

/**
 * Pagination info
 */
@Serializable
data class PaginationInfo(
    val page: Int,
    val limit: Int,
    val hasNext: Boolean,
    val hasPrevious: Boolean
)

/**
 * API key validation response
 */
@Serializable
data class APIKeyValidationResponse(
    val isValid: Boolean,
    val permissions: List<String>,
    @Serializable(with = DateSerializer::class)
    val expiresAt: Date?,
    val rateLimit: RateLimitInfo
)

/**
 * Rate limit info
 */
@Serializable
data class RateLimitInfo(
    val requestsPerMinute: Int,
    val requestsRemaining: Int,
    @Serializable(with = DateSerializer::class)
    val resetTime: Date
)

/**
 * Usage statistics
 */
@Serializable
data class UsageStatistics(
    val totalAnalyses: Int,
    val currentMonthAnalyses: Int,
    val averageConfidence: Double,
    val topKeywords: List<String>,
    val documentTypes: Map<String, Int>,
    val successRate: Double
)

// Serializers for custom types

@kotlinx.serialization.Serializer(forClass = Duration::class)
object DurationSerializer : kotlinx.serialization.KSerializer<Duration> {
    override val descriptor = kotlinx.serialization.descriptors.PrimitiveSerialDescriptor("Duration", kotlinx.serialization.descriptors.PrimitiveKind.LONG)
    
    override fun serialize(encoder: kotlinx.serialization.encoding.Encoder, value: Duration) {
        encoder.encodeLong(value.inWholeMilliseconds)
    }
    
    override fun deserialize(decoder: kotlinx.serialization.encoding.Decoder): Duration {
        return Duration.ofMillis(decoder.decodeLong())
    }
}

@kotlinx.serialization.Serializer(forClass = Date::class)
object DateSerializer : kotlinx.serialization.KSerializer<Date> {
    override val descriptor = kotlinx.serialization.descriptors.PrimitiveSerialDescriptor("Date", kotlinx.serialization.descriptors.PrimitiveKind.LONG)
    
    override fun serialize(encoder: kotlinx.serialization.encoding.Encoder, value: Date) {
        encoder.encodeLong(value.time)
    }
    
    override fun deserialize(decoder: kotlinx.serialization.encoding.Decoder): Date {
        return Date(decoder.decodeLong())
    }
}

File: sdks/ios/Sources/ArbitrationSDK/Models/ArbitrationModels.swift
import Foundation
import UIKit

// MARK: - Analysis Result
public struct ArbitrationAnalysisResult: Codable, Identifiable {
    public let id: String
    public let hasArbitration: Bool
    public let confidence: Double
    public let keywordMatches: [KeywordMatch]
    public let exclusionMatches: [KeywordMatch]
    public let riskLevel: RiskLevel
    public let recommendations: [String]
    public let metadata: AnalysisMetadata
    public var processingTime: TimeInterval = 0
    
    public init(
        id: String,
        hasArbitration: Bool,
        confidence: Double,
        keywordMatches: [KeywordMatch],
        exclusionMatches: [KeywordMatch],
        riskLevel: RiskLevel,
        recommendations: [String],
        metadata: AnalysisMetadata
    ) {
        self.id = id
        self.hasArbitration = hasArbitration
        self.confidence = confidence
        self.keywordMatches = keywordMatches
        self.exclusionMatches = exclusionMatches
        self.riskLevel = riskLevel
        self.recommendations = recommendations
        self.metadata = metadata
    }
}

// MARK: - Keyword Match
public struct KeywordMatch: Codable, Identifiable {
    public let id = UUID()
    public let keyword: String
    public let range: String // Stored as string for Codable
    public let context: String
    public let confidence: Double
    
    public init(keyword: String, range: Range<String.Index>, context: String, confidence: Double) {
        self.keyword = keyword
        self.range = "\(range.lowerBound.utf16Offset(in: context)):\(range.upperBound.utf16Offset(in: context))"
        self.context = context
        self.confidence = confidence
    }
    
    private enum CodingKeys: String, CodingKey {
        case keyword, range, context, confidence
    }
}

// MARK: - Risk Level
public enum RiskLevel: String, Codable, CaseIterable {
    case minimal = "minimal"
    case low = "low"
    case medium = "medium"
    case high = "high"
    
    public var color: UIColor {
        switch self {
        case .minimal:
            return .systemGreen
        case .low:
            return .systemYellow
        case .medium:
            return .systemOrange
        case .high:
            return .systemRed
        }
    }
    
    public var description: String {
        switch self {
        case .minimal:
            return "Minimal Risk"
        case .low:
            return "Low Risk"
        case .medium:
            return "Medium Risk"
        case .high:
            return "High Risk"
        }
    }
}

// MARK: - Analysis Metadata
public struct AnalysisMetadata: Codable {
    public let textLength: Int
    public let processingTime: TimeInterval
    public let sdkVersion: String
    public let modelVersion: String
    public let analysisDate: Date
    
    public init(
        textLength: Int,
        processingTime: TimeInterval,
        sdkVersion: String,
        modelVersion: String,
        analysisDate: Date
    ) {
        self.textLength = textLength
        self.processingTime = processingTime
        self.sdkVersion = sdkVersion
        self.modelVersion = modelVersion
        self.analysisDate = analysisDate
    }
}

// MARK: - Rule Based Result
public struct RuleBasedResult {
    public let hasArbitration: Bool
    public let confidence: Double
    public let keywordMatches: [KeywordMatch]
    public let exclusionMatches: [KeywordMatch]
    
    public init(
        hasArbitration: Bool,
        confidence: Double,
        keywordMatches: [KeywordMatch],
        exclusionMatches: [KeywordMatch]
    ) {
        self.hasArbitration = hasArbitration
        self.confidence = confidence
        self.keywordMatches = keywordMatches
        self.exclusionMatches = exclusionMatches
    }
}

// MARK: - Document Type
public enum DocumentType: String, Codable, CaseIterable {
    case pdf = "pdf"
    case word = "word"
    case text = "text"
    case rtf = "rtf"
    case html = "html"
    case image = "image"
    
    public var supportedExtensions: [String] {
        switch self {
        case .pdf:
            return ["pdf"]
        case .word:
            return ["doc", "docx"]
        case .text:
            return ["txt"]
        case .rtf:
            return ["rtf"]
        case .html:
            return ["html", "htm"]
        case .image:
            return ["jpg", "jpeg", "png", "tiff", "bmp"]
        }
    }
}

// MARK: - SDK Configuration
public struct SDKConfiguration {
    public let apiBaseURL: String
    public let apiKey: String
    public let confidenceThreshold: Double
    public let enableOfflineMode: Bool
    public let enableAnalytics: Bool
    public let cacheSize: Int
    public let networkTimeout: TimeInterval
    
    public init(
        apiBaseURL: String = "https://api.arbitration-platform.com",
        apiKey: String = "",
        confidenceThreshold: Double = 0.5,
        enableOfflineMode: Bool = true,
        enableAnalytics: Bool = true,
        cacheSize: Int = 100,
        networkTimeout: TimeInterval = 30
    ) {
        self.apiBaseURL = apiBaseURL
        self.apiKey = apiKey
        self.confidenceThreshold = confidenceThreshold
        self.enableOfflineMode = enableOfflineMode
        self.enableAnalytics = enableAnalytics
        self.cacheSize = cacheSize
        self.networkTimeout = networkTimeout
    }
    
    public static let `default` = SDKConfiguration()
}

// MARK: - Arbitration Error
public enum ArbitrationError: LocalizedError {
    case invalidInput(String)
    case analysisFailure(String)
    case networkError(String)
    case storageError(String)
    case unsupportedDocumentType(String)
    case authenticationFailure(String)
    case rateLimitExceeded
    case modelNotAvailable
    
    public var errorDescription: String? {
        switch self {
        case .invalidInput(let message):
            return "Invalid input: \(message)"
        case .analysisFailure(let message):
            return "Analysis failed: \(message)"
        case .networkError(let message):
            return "Network error: \(message)"
        case .storageError(let message):
            return "Storage error: \(message)"
        case .unsupportedDocumentType(let message):
            return "Unsupported document type: \(message)"
        case .authenticationFailure(let message):
            return "Authentication failed: \(message)"
        case .rateLimitExceeded:
            return "Rate limit exceeded. Please try again later."
        case .modelNotAvailable:
            return "ML model is not available. Using fallback detection."
        }
    }
}

// MARK: - Analysis Request
public struct AnalysisRequest: Codable {
    public let text: String
    public let requestId: String
    public let options: AnalysisOptions
    
    public init(text: String, requestId: String = UUID().uuidString, options: AnalysisOptions = .default) {
        self.text = text
        self.requestId = requestId
        self.options = options
    }
}

// MARK: - Analysis Options
public struct AnalysisOptions: Codable {
    public let includeConfidenceScore: Bool
    public let includeKeywordMatches: Bool
    public let includeRecommendations: Bool
    public let language: String
    public let strictMode: Bool
    
    public init(
        includeConfidenceScore: Bool = true,
        includeKeywordMatches: Bool = true,
        includeRecommendations: Bool = true,
        language: String = "en",
        strictMode: Bool = false
    ) {
        self.includeConfidenceScore = includeConfidenceScore
        self.includeKeywordMatches = includeKeywordMatches
        self.includeRecommendations = includeRecommendations
        self.language = language
        self.strictMode = strictMode
    }
    
    public static let `default` = AnalysisOptions()
}

// MARK: - Batch Analysis Result
public struct BatchAnalysisResult: Codable {
    public let batchId: String
    public let results: [ArbitrationAnalysisResult]
    public let summary: BatchSummary
    public let processingTime: TimeInterval
    
    public init(batchId: String, results: [ArbitrationAnalysisResult], summary: BatchSummary, processingTime: TimeInterval) {
        self.batchId = batchId
        self.results = results
        self.summary = summary
        self.processingTime = processingTime
    }
}

// MARK: - Batch Summary
public struct BatchSummary: Codable {
    public let totalDocuments: Int
    public let documentsWithArbitration: Int
    public let averageConfidence: Double
    public let highRiskDocuments: Int
    public let processingErrors: Int
    
    public init(
        totalDocuments: Int,
        documentsWithArbitration: Int,
        averageConfidence: Double,
        highRiskDocuments: Int,
        processingErrors: Int
    ) {
        self.totalDocuments = totalDocuments
        self.documentsWithArbitration = documentsWithArbitration
        self.averageConfidence = averageConfidence
        self.highRiskDocuments = highRiskDocuments
        self.processingErrors = processingErrors
    }
}

// MARK: - Analytics Event
public struct AnalyticsEvent: Codable {
    public let eventType: AnalyticsEventType
    public let timestamp: Date
    public let properties: [String: String]
    
    public init(eventType: AnalyticsEventType, properties: [String: String] = [:]) {
        self.eventType = eventType
        self.timestamp = Date()
        self.properties = properties
    }
}

// MARK: - Analytics Event Type
public enum AnalyticsEventType: String, Codable {
    case analysisStarted = "analysis_started"
    case analysisCompleted = "analysis_completed"
    case analysisFailed = "analysis_failed"
    case documentScanned = "document_scanned"
    case offlineModeEnabled = "offline_mode_enabled"
    case cacheHit = "cache_hit"
    case apiCallMade = "api_call_made"
}

// MARK: - Notification Settings
public struct NotificationSettings: Codable {
    public let enablePushNotifications: Bool
    public let enableAnalysisComplete: Bool
    public let enableHighRiskAlerts: Bool
    public let enableWeeklyReports: Bool
    
    public init(
        enablePushNotifications: Bool = true,
        enableAnalysisComplete: Bool = true,
        enableHighRiskAlerts: Bool = true,
        enableWeeklyReports: Bool = false
    ) {
        self.enablePushNotifications = enablePushNotifications
        self.enableAnalysisComplete = enableAnalysisComplete
        self.enableHighRiskAlerts = enableHighRiskAlerts
        self.enableWeeklyReports = enableWeeklyReports
    }
}

// MARK: - User Preferences
public struct UserPreferences: Codable {
    public let language: String
    public let theme: AppTheme
    public let notifications: NotificationSettings
    public let autoSave: Bool
    public let biometricAuth: Bool
    
    public init(
        language: String = "en",
        theme: AppTheme = .system,
        notifications: NotificationSettings = NotificationSettings(),
        autoSave: Bool = true,
        biometricAuth: Bool = false
    ) {
        self.language = language
        self.theme = theme
        self.notifications = notifications
        self.autoSave = autoSave
        self.biometricAuth = biometricAuth
    }
}

// MARK: - App Theme
public enum AppTheme: String, Codable, CaseIterable {
    case light = "light"
    case dark = "dark"
    case system = "system"
    
    public var displayName: String {
        switch self {
        case .light:
            return "Light"
        case .dark:
            return "Dark"
        case .system:
            return "System"
        }
    }
}